### DISCUSSION TITLE
Does poetry support clean unnecessary dependency specification in pyproject.toml?

### DISCUSSION THREAD

**Discussion by dclong** [python-poetry/poetry#3507] [Q&A]
Assume there's a dependency A which depends on B. If a specification is specified on A and another specification is claimed on B but doesn't add any more constraints in addition to the specification on A. Is there any way to clean the unnecessary specification on B?

**Comment by sinoroc** [comment]
I do not follow exactly what you mean... Maybe you could show with more concrete examples, excerpts of `pyproject.toml` files, and/or poetry commands.

One thing: you should declare all your direct dependencies in `pyproject.toml` and only your direct dependencies. If you depend on both A and B, then you have to declare them both. It does not matter if A already depends on B, this is not your concern. It could be that in a next version of A, A drops its dependency on B, but your project still depends on B.

**Comment by ricardo-reis-1970** [comment]
I too have this question, but I'll try and put it in a clear way, as opposed to A and B. Say I have this project and a number of dependencies were added to it that are just not needed. They are not indirect dependencies or anything.

Concretely: I added `openCV` , `open3d` and `pandas` to a main project but later decided that it would be better to factor out the parts dealing with Computer Vision into a separate project and just _optionally_ include that in the main project, depending on deployment. Now, my main project is including these libraries for nothing other than filling limited disk resources.

You might be thinking: "Well, he _knows_ what to delete. Why doesn't he just go there and delete it, then?" You're right, in a limited fashion: I know what to delete in one project. Lots of dependencies have been added just for trial and then left there hanging in as many as 30 different projects and sub projects. A mess.

`Pycln` seems to do this, but only for the actual python files. What I need is a tool that looks at the `pyproject.toml` and the whole project and decides that this dependency is no longer needed.

We all have an option, but for some reason no one seems to like it:
1. destroy the virtual environment;
2. remove one dependency
3. recreate the whole thing

If we iterate over all dependencies, we'll eventually come across those that actually cause no harm. Well, except when someone decided that importing inside `if` clauses is a jolly good idea. Now, imagine you have 45 dependencies...

Is there a better solution for this? Please note that _any_ solution is a better solution than the brute force one outlined above.

**Reply by sinoroc** [reply]
I do not know of anything specific to poetry. Some ideas:
* https://stackoverflow.com/a/63420404
* https://stackoverflow.com/a/61540466
* https://stackoverflow.com/a/61202584
* https://stackoverflow.com/a/61025535

**Reply by finswimmer** [reply]
I would just remove all dependency and ask my idea (PyCharm) to show me unresolved references. Than I can add the missing dependencies one by one.
